### PR TITLE
[Global Styles]: Add search in block types list

### LINKED ISSUE #39135: Allow block search option on editor for global styles
## What problem does this address?
Currently, We have to scroll and find the specific block from the list of all the available blocks. There are no search options for a long block list.

https://user-images.githubusercontent.com/25276905/156142717-2f0f17ef-d67d-4514-ac8b-c870448d5add.mov

## What is your proposed solution?
There should be a search/filter option where users can type and search/filter the blocks. This will reduce the block search time it takes when having a long list of blocks.

### PR DESCRIPTION
**Pull request by ntsekouras**
Resolves: https://github.com/WordPress/gutenberg/issues/39135

This PR implements search in Global Styles block types list.

## Testing instructions
1. with a block theme go to site editor and open the Global Styles panel
2. select the `blocks` to open the list of them
3. search blocks.

https://user-images.githubusercontent.com/16275880/155954806-ced79517-375f-481a-9186-cf0af180540c.mov

## Notes
1. If you search and select a block and then go back to the list the search term is not preserved. This is because the the component unmounts and in order to preserve this we should create a new state prop. I don't think the alternative of lifting the state up or making changes in navigation components worths it. I believe this is okay for at least this first iteration..
3. For announcing the results I'm using `childElementCount` from the underlying DOM element wrapper for the block type items. I do this because filtered items can contain items that will eventually not render and there is no reliable way to detect when a child will return `null`.

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice, seems good to me as the block list can become very long! Reused search box from preferences works okay here.

File: packages/block-editor/src/components/index.js [hunk 1]
```diff
@@ -126,6 +126,7 @@ export {
 export { default as DefaultBlockAppender } from './default-block-appender';
 export { default as __unstableEditorStyles } from './editor-styles';
 export { default as Inserter } from './inserter';
+export { searchItems } from './inserter/search-items';
```
Comment (youknowriad) [inline packages/block-editor/src/components/index.js]: I think "searchItems" here is confusing and I actually wonder if it should be used in this PR. Basically this function is used to "search inserter items" which are not exactly block types:

 - A reusable block type with initial attributes is a search item
 - A block variation is a search item
 - A block type is a search item
 - I wonder if a pattern is a search item or if it has a different search function.

In other words, it seems too specific to the inserter for me, maybe we could have a "searchBlockType" function or something like that, I also don't see why it should be in "block-editor" instead of being a "blocks" thing. (The inserter search makes sense in block-editor though)

Comment (youknowriad) [reply packages/block-editor/src/components/index.js]: It can also be a selector in `core/blocks` package (we do have selectors to fetch block types there)

Comment (ntsekouras) [reply packages/block-editor/src/components/index.js]: >I think "searchItems" here is confusing and I actually wonder if it should be used in this PR.

Agreed! A selector might be better here and I think it already exists... I'll update soon.

>I wonder if a pattern is a search item or if it has a different search function.

For the record it's used for patterns in the inserter too.

Comment (ntsekouras) [reply packages/block-editor/src/components/index.js]: Updated to use `isMatchingSearchTerm` selector. Thanks Riad for the feedback!

File: packages/edit-site/src/components/global-styles/screen-block-list.js [hunk 2]
```diff
@@ -45,6 +50,44 @@ function BlockMenuItem( { block } ) {
...
+	// Announce search results on change
+	useEffect( () => {
+		if ( ! filterValue ) {
+			return;
+		}
+		// We extract the results from the wrapper div's `ref` because
+		// filtered items can contain items that will eventually not
+		// render and there is no reliable way to detect when a child
+		// will return `null`.
+		const count = blockTypesListRef.current.childElementCount;
```
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/screen-block-list.js:105]: This is very weird 😬

Comment (youknowriad) [reply packages/edit-site/src/components/global-styles/screen-block-list.js:105]: I can we still try to count using data (filter the results by checking if the block types actually have global styles panels, but not sure how complex would that be)

**Review by youknowriad — APPROVED** [review]
The way to get the count is very weird IMO and subject to easy breakage (imagine the component start to render two elements instead of one)

File: packages/edit-site/src/components/global-styles/screen-block-list.js (hunk 2 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/global-styles/screen-block-list.js:105]: >but not sure how complex would that be

I think it's very complex and involves many hooks with lots of logic and things like mapping props ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/global-styles/hooks.js#L171)), etc..

I agree it's very fragile but couldn't find a better solution for this without porting all the logic of hooks in selectors.

I added a comment about this to be easier to trace if this happens in the future..  Who knows, maybe React will support this info in a future version 😄

Comment (mtias) [reply packages/edit-site/src/components/global-styles/screen-block-list.js:105]: I think it's fine to list blocks that don't have global styles properties in search results and if you select one it'd say "no style properties available" or something

Comment (ntsekouras) [reply packages/edit-site/src/components/global-styles/screen-block-list.js:105]: The number can differ a lot. For example without other plugins enabled the block types are `90`, but the ones with global style properties are `52`. While searching, the numbers will be smaller, but still it can result in many 'no need to be there' block types.

**Comment by jasmussen** [comment]
I found myself missing this one a lot today!

**Comment by ntsekouras** [comment]
I think we can land this as a stopgap solution and I added an e2e test to make sure we catch the case if the number of `BlockMenuItem`'s rendered elements  changes. --cc @youknowriad

**Comment by youknowriad** [comment]
Makes sense to me. Thanks Nik

**Comment by ntsekouras** [comment]
@gziolo , @adamziel is this something that could be added for 6.0 at this point?

**Comment by gziolo** [comment]
This PR was cherry-picked for the Gutenberg plugin v13.0 release and WordPress 6.0 RC1 release.
